### PR TITLE
Fix random CI failures in globanet jobs

### PR DESCRIPTION
**Pull request by sridhargaddam**
This PR reorganizes the code slightly and includes additional
logging which will assist whenever there is any failure in GN
jobs.

Partially fixes: https://github.com/submariner-io/submariner/issues/596

Signed-off-by: Sridhar Gaddam <sgaddam@redhat.com>

### CODE REVIEW

**Comment by mangelajo** [comment]
@sridhargaddam can include the go.mod update to the new shipyard patches?
```
GO111MODULE=on go get github.com/submariner-io/shipyard@master
 ```

I think it is ^

File: pkg/globalnet/controllers/ipam/ipam.go [hunk 1]
```diff
@@ -259,33 +270,44 @@ func (i *Controller) handleUpdatePod(old interface{}, newObj interface{}) {
...
+		klog.V(log.DEBUG).Infof("Pod %s with ip %s is being terminated", key, oldPodIp)
 		i.handleRemovedPod(old)
+		return
 	}
 
-	// Ignore privileged pods that use hostNetwork
-	if updatedPodIp != "" && updatedPodIp == podHostIP {
-		klog.V(log.DEBUG).Infof("Pod %s on host %s uses hostNetwork, ignoring", old.(*k8sv1.Pod).Name, podHostIP)
-		return
+	newGlobalIp := newObj.(*k8sv1.Pod).GetAnnotations()[submarinerIpamGlobalIp]
```
[the remark this answers is not in the document]
Comment (mangelajo) [reply pkg/globalnet/controllers/ipam/ipam.go:293]: Oh you're right, sorry, I thought a golang map with unknown key panicked... 

still learning go it seems ':)

File: pkg/globalnet/controllers/ipam/gatewaymonitor.go [hunk 2]
```diff
@@ -77,7 +77,7 @@ func NewGatewayMonitor(spec *SubmarinerIpamControllerSpecification, cfg *rest.Co
 
 func (i *GatewayMonitor) Run(stopCh <-chan struct{}) error {
 	defer utilruntime.HandleCrash()
-	defer ClearGlobalNetChains(i.ipt)
+	//defer ClearGlobalNetChains(i.ipt)
```
Comment (mangelajo) [inline pkg/globalnet/controllers/ipam/gatewaymonitor.go:80]:
```suggestion
```

File: pkg/globalnet/controllers/ipam/ipam.go [hunk 3]
```diff
@@ -177,8 +177,8 @@ func (i *Controller) processNextObject(objWorkqueue workqueue.RateLimitingInterf
 				}
 
 				// Privileged pods that use hostNetwork will be ignored.
-				if runtimeObj.Status.PodIP == runtimeObj.Status.HostIP {
-					klog.V(log.DEBUG).Infof("Ignoring pod %s on host %s as it uses hostNetworking", runtimeObj.Name, runtimeObj.Status.PodIP)
+				if runtimeObj.Spec.HostNetwork {
+					klog.V(log.DEBUG).Infof("Ignoring pod %s on host %s as it uses hostNetworking", key, runtimeObj.Status.PodIP)
```
Comment (mangelajo) [inline pkg/globalnet/controllers/ipam/ipam.go:181]:
```suggestion
					klog.V(log.DEBUG).Infof("Ignoring pod %q on host %q as it uses hostNetworking", key, runtimeObj.Status.PodIP)
```

while we are modifying the code.. :) so we are uniform with that in other places.. @tpantelis convinced me.. ':)

File: pkg/globalnet/controllers/ipam/ipam.go [hunk 4]
```diff
@@ -243,11 +243,22 @@ func (i *Controller) handleUpdateService(old interface{}, newObj interface{}) {
 		utilruntime.HandleError(err)
 		return
 	}
+
+	if i.excludeNamespaces[newObj.(*k8sv1.Service).Namespace] {
+		klog.V(log.DEBUG).Infof("In handleUpdateService, skipping Service %s as it belongs to excluded namespace.", key)
```
Comment (mangelajo) [inline pkg/globalnet/controllers/ipam/ipam.go:248]:
```suggestion
		klog.V(log.DEBUG).Infof("In handleUpdateService, skipping Service %q as it belongs to excluded namespace.", key)
```

File: pkg/globalnet/controllers/ipam/globalnet.go [hunk 5]
```diff
@@ -92,6 +92,8 @@ func (i *Controller) evaluateService(service *k8sv1.Service) Operation {
 	if chainExists, _ := i.doesIPTablesChainExist("nat", chainName); !chainExists {
 		return Requeue
 	}
+	serviceName := service.GetNamespace() + "/" + service.GetName()
+	klog.V(log.DEBUG).Infof("kube-proxy chain %s for service %s now exists.", chainName, serviceName)
```
Comment (mangelajo) [inline pkg/globalnet/controllers/ipam/globalnet.go:96]:
```suggestion
	klog.V(log.DEBUG).Infof("kube-proxy chain %q for service %q now exists.", chainName, serviceName)
```

File: pkg/globalnet/controllers/ipam/ipam.go [hunk 6]
```diff
@@ -243,11 +243,22 @@ func (i *Controller) handleUpdateService(old interface{}, newObj interface{}) {
...
+		return
+	}
+
 	oldGlobalIp := old.(*k8sv1.Service).GetAnnotations()[submarinerIpamGlobalIp]
 	newGlobalIp := newObj.(*k8sv1.Service).GetAnnotations()[submarinerIpamGlobalIp]
 	if oldGlobalIp != newGlobalIp && newGlobalIp != i.pool.GetAllocatedIp(key) {
-		klog.V(log.DEBUG).Infof("GlobalIp changed from %s to %s for %s", oldGlobalIp, newGlobalIp, key)
-		i.enqueueObject(newObj, i.serviceWorkqueue)
+		klog.V(log.DEBUG).Infof("GlobalIp changed from %s to %s for Service %s", oldGlobalIp, newGlobalIp, key)
```
Comment (mangelajo) [inline pkg/globalnet/controllers/ipam/ipam.go:255]:
```suggestion
		klog.V(log.DEBUG).Infof("GlobalIp changed from %s to %s for Service %q", oldGlobalIp, newGlobalIp, key)
```

File: pkg/globalnet/controllers/ipam/ipam.go [hunk 7]
```diff
@@ -259,33 +270,44 @@ func (i *Controller) handleUpdatePod(old interface{}, newObj interface{}) {
...
 
-	oldPodIp := old.(*k8sv1.Pod).Status.PodIP
-	updatedPodIp := newObj.(*k8sv1.Pod).Status.PodIP
-	podHostIP := newObj.(*k8sv1.Pod).Status.HostIP
-	// Pod events that are skipped during addEvent are handled here when they are assigned an ipaddress.
-	if updatedPodIp != "" && oldPodIp != updatedPodIp {
-		klog.V(log.DEBUG).Infof("In handleUpdatePod, pod %s is now assigned %s address, enqueing", old.(*k8sv1.Pod).Name, updatedPodIp)
-		i.enqueueObject(newObj, i.podWorkqueue)
+	if i.excludeNamespaces[newObj.(*k8sv1.Pod).Namespace] {
+		klog.V(log.DEBUG).Infof("In handleUpdatePod, skipping pod %s as it belongs to excluded namespace.", key)
```
Comment (mangelajo) [inline pkg/globalnet/controllers/ipam/ipam.go:274]:
```suggestion
		klog.V(log.DEBUG).Infof("In handleUpdatePod, skipping pod %q as it belongs to excluded namespace.", key)
```

File: pkg/globalnet/controllers/ipam/ipam.go [hunk 8]
```diff
@@ -259,33 +270,44 @@ func (i *Controller) handleUpdatePod(old interface{}, newObj interface{}) {
...
 		return
 	}
 
+	// Ignore privileged pods that use hostNetwork
+	if newObj.(*k8sv1.Pod).Spec.HostNetwork {
+		klog.V(log.DEBUG).Infof("Pod %s on host %s uses hostNetwork, ignoring", key, newObj.(*k8sv1.Pod).Status.HostIP)
```
Comment (mangelajo) [inline pkg/globalnet/controllers/ipam/ipam.go:280]:
```suggestion
		klog.V(log.DEBUG).Infof("Pod %q on host %q uses hostNetwork, ignoring", key, newObj.(*k8sv1.Pod).Status.HostIP)
```

File: pkg/globalnet/controllers/ipam/ipam.go [hunk 9]
```diff
@@ -259,33 +270,44 @@ func (i *Controller) handleUpdatePod(old interface{}, newObj interface{}) {
...
+		return
+	}
+
+	oldPodIp := old.(*k8sv1.Pod).Status.PodIP
+	updatedPodIp := newObj.(*k8sv1.Pod).Status.PodIP
 	// When the POD is getting terminated, sometimes we get pod update event with podIp removed.
 	if oldPodIp != "" && updatedPodIp == "" {
-		klog.V(log.DEBUG).Infof("Pod %s with ip %s is being terminated", old.(*k8sv1.Pod).Name, oldPodIp)
+		klog.V(log.DEBUG).Infof("Pod %s with ip %s is being terminated", key, oldPodIp)
```
Comment (mangelajo) [inline pkg/globalnet/controllers/ipam/ipam.go:288]:
```suggestion
		klog.V(log.DEBUG).Infof("Pod %q with ip %s is being terminated", key, oldPodIp)
```

File: pkg/globalnet/controllers/ipam/ipam.go [hunk 10]
```diff
@@ -259,33 +270,44 @@ func (i *Controller) handleUpdatePod(old interface{}, newObj interface{}) {
...
 
-	// Ignore privileged pods that use hostNetwork
-	if updatedPodIp != "" && updatedPodIp == podHostIP {
-		klog.V(log.DEBUG).Infof("Pod %s on host %s uses hostNetwork, ignoring", old.(*k8sv1.Pod).Name, podHostIP)
-		return
+	newGlobalIp := newObj.(*k8sv1.Pod).GetAnnotations()[submarinerIpamGlobalIp]
+	if newGlobalIp == "" {
+		// Pod events that are skipped during addEvent are handled here when they are assigned an ipaddress.
+		if updatedPodIp != "" {
+			klog.V(log.DEBUG).Infof("In handleUpdatePod, pod %s is now assigned %s address, enqueing", key, updatedPodIp)
```
Comment (mangelajo) [inline pkg/globalnet/controllers/ipam/ipam.go:297]:
```suggestion
			klog.V(log.DEBUG).Infof("In handleUpdatePod, pod %q is now assigned %s address, enqueing", key, updatedPodIp)
```

File: pkg/globalnet/controllers/ipam/ipam.go [hunk 11]
```diff
@@ -349,8 +370,10 @@ func (i *Controller) handleRemovedService(obj interface{}) {
 			klog.V(log.DEBUG).Infof("Released ip %s for service %s", globalIp, key)
 			err = i.syncServiceRules(service, globalIp, DeleteRules)
 			if err != nil {
-				klog.Errorf("Error while cleaning up Service ingress rules. %v", err)
+				klog.Errorf("Error while cleaning up Service %s ingress rules. %v", key, err)
```
Comment (mangelajo) [inline pkg/globalnet/controllers/ipam/ipam.go:373]:
```suggestion
				klog.Errorf("Error while cleaning up Service %q ingress rules. %+v", key, err)
```

File: pkg/globalnet/controllers/ipam/ipam.go [hunk 12]
```diff
@@ -349,8 +370,10 @@ func (i *Controller) handleRemovedService(obj interface{}) {
...
 			}
+		} else {
+			klog.Errorf("Error: handleRemovedService called for %s, but globalIp annotation is missing.", key)
```
Comment (mangelajo) [inline pkg/globalnet/controllers/ipam/ipam.go:376]:
```suggestion
			klog.Errorf("Error: handleRemovedService called for %q, but globalIp annotation is missing.", key)
```

File: pkg/globalnet/controllers/ipam/ipam.go [hunk 13]
```diff
@@ -373,19 +396,21 @@ func (i *Controller) handleRemovedPod(obj interface{}) {
...
-		klog.V(log.TRACE).Infof("handleRemovedPod called for Pod %s", key)
 		globalIp := pod.Annotations[submarinerIpamGlobalIp]
 		if globalIp != "" && pod.Status.PodIP != "" {
 			if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
 				utilruntime.HandleError(err)
 				return
 			}
 			i.pool.Release(key)
-			klog.V(log.DEBUG).Infof("Released GlobalIp %s for pod %s", globalIp, key)
+			klog.V(log.DEBUG).Infof("Released globalIp %s for pod %s", globalIp, key)
```
Comment (mangelajo) [inline pkg/globalnet/controllers/ipam/ipam.go:406]:
```suggestion
			klog.V(log.DEBUG).Infof("Released globalIp %s for pod %s", globalIp, key)
```
```suggestion
			klog.V(log.DEBUG).Infof("Released globalIp %s for pod %q", globalIp, key)
```

File: pkg/globalnet/controllers/ipam/ipam.go [hunk 14]
```diff
@@ -373,19 +396,21 @@ func (i *Controller) handleRemovedPod(obj interface{}) {
...
 			err = i.syncPodRules(pod.Status.PodIP, globalIp, DeleteRules)
 			if err != nil {
 				klog.Errorf("Error while cleaning up Pod egress rules. %v", err)
 			}
+		} else {
+			podKey := pod.Namespace + "/" + pod.Name
+			klog.V(log.DEBUG).Infof("handleRemovedPod called for %s, that has globalIp %q and PodIp %q", podKey, globalIp, pod.Status.PodIP)
```
Comment (mangelajo) [inline pkg/globalnet/controllers/ipam/ipam.go:413]:
```suggestion
			klog.V(log.DEBUG).Infof("handleRemovedPod called for %q, that has globalIp %v and PodIp %v", podKey, globalIp, pod.Status.PodIP)
```

File: pkg/globalnet/controllers/ipam/ipam.go [hunk 15]
```diff
@@ -427,6 +452,8 @@ func (i *Controller) handleRemovedNode(obj interface{}) {
 		if err != nil {
 			klog.Errorf("Error while cleaning up HostNetwork egress rules. %v", err)
 		}
+	} else {
+		klog.V(log.DEBUG).Infof("handleRemovedNode called for %s, that has globalIp %q and cniIfaceIp %q", key, globalIp, cniIfaceIp)
```
Comment (mangelajo) [inline pkg/globalnet/controllers/ipam/ipam.go:456]:
```suggestion
		klog.V(log.DEBUG).Infof("handleRemovedNode called for %q, that has globalIp %s and cniIfaceIp %s", key, globalIp, cniIfaceIp)
```

File: pkg/globalnet/controllers/ipam/ipam.go [hunk 16]
```diff
@@ -447,7 +473,7 @@ func (i *Controller) annotateGlobalIp(key string, globalIp string) (string, erro
 	}
 	if globalIp != givenIp {
 		// This resource has been allocated a different IP
-		klog.Warningf("Updating GlobalIp for %s from %s to %s", key, globalIp, givenIp)
+		klog.Warningf("Updating globalIp for %s from %s to %s", key, globalIp, givenIp)
```
Comment (mangelajo) [inline pkg/globalnet/controllers/ipam/ipam.go:476]:
```suggestion
		klog.Warningf("Updating globalIp for %q from %s to %s", key, globalIp, givenIp)
```

File: pkg/globalnet/controllers/ipam/ipam.go [hunk 17]
```diff
@@ -480,6 +506,7 @@ func (i *Controller) serviceUpdater(obj runtime.Object, key string) error {
 	// 1. when the Service does not have the globalIp annotation and a new globalIp is allocated
 	// 2. when the current globalIp annotation on the Service does not match with the info maintained by ipPool
 	if allocatedIp != "" {
+		klog.V(log.DEBUG).Infof("Allocating globalIp %s to Service %s ", allocatedIp, key)
```
Comment (mangelajo) [inline pkg/globalnet/controllers/ipam/ipam.go:509]:
```suggestion
		klog.V(log.DEBUG).Infof("Allocating globalIp %s to Service %q ", allocatedIp, key)
```

File: pkg/globalnet/controllers/ipam/ipam.go [hunk 18]
```diff
@@ -498,7 +525,7 @@ func (i *Controller) serviceUpdater(obj runtime.Object, key string) error {
 			return err
 		}
 	} else if existingGlobalIp != "" {
-		klog.V(log.DEBUG).Infof("GatewayNode seems to have migrated, sync rules for Service %q", service.Name)
+		klog.V(log.DEBUG).Infof("Service %q already has globalIp %q annotation, syncing rules", key, existingGlobalIp)
```
Comment (mangelajo) [inline pkg/globalnet/controllers/ipam/ipam.go:528]:
```suggestion
		klog.V(log.DEBUG).Infof("Service %q already has globalIp %s annotation, syncing rules", key, existingGlobalIp)
```

File: pkg/globalnet/controllers/ipam/ipam.go [hunk 19]
```diff
@@ -520,13 +547,14 @@ func (i *Controller) podUpdater(obj runtime.Object, key string) error {
 	allocatedIp, err := i.annotateGlobalIp(key, existingGlobalIp)
 	if err != nil { // failed to get globalIp or failed to update, we want to retry
 		logAndRequeue(key, i.podWorkqueue)
-		return fmt.Errorf("failed to annotate GlobalIp to Pod %s: %v", key, err)
+		return fmt.Errorf("failed to annotate globalIp to Pod %s: %v", key, err)
```
Comment (mangelajo) [inline pkg/globalnet/controllers/ipam/ipam.go:550]:
```suggestion
		return fmt.Errorf("failed to annotate globalIp to Pod %q: %+v", key, err)
```

File: pkg/globalnet/controllers/ipam/ipam.go [hunk 20]
```diff
@@ -520,13 +547,14 @@ func (i *Controller) podUpdater(obj runtime.Object, key string) error {
...
 	}
 
 	// This case is hit in one of the two situations
 	// 1. when the POD does not have the globalIp annotation and a new globalIp is allocated
 	// 2. when the current globalIp annotation on the POD does not match with the info maintained by ipPool
 	if allocatedIp != "" {
+		klog.V(log.DEBUG).Infof("Allocating globalIp %s to Pod %s ", allocatedIp, key)
```
Comment (mangelajo) [inline pkg/globalnet/controllers/ipam/ipam.go:557]:
```suggestion
		klog.V(log.DEBUG).Infof("Allocating globalIp %s to Pod %q ", allocatedIp, key)
```

File: pkg/globalnet/controllers/ipam/ipam.go [hunk 21]
```diff
@@ -545,7 +573,7 @@ func (i *Controller) podUpdater(obj runtime.Object, key string) error {
 			return err
 		}
 	} else if existingGlobalIp != "" {
-		klog.V(log.DEBUG).Infof("GatewayNode seems to have migrated, sync rules for Pod %q", pod.Name)
+		klog.V(log.DEBUG).Infof("Pod %q already has globalIp %q annotation, syncing rules", key, existingGlobalIp)
```
Comment (mangelajo) [inline pkg/globalnet/controllers/ipam/ipam.go:576]:
```suggestion
		klog.V(log.DEBUG).Infof("Pod %q already has globalIp %s annotation, syncing rules", key, existingGlobalIp)
```

File: pkg/globalnet/controllers/ipam/ipam.go [hunk 22]
```diff
@@ -566,13 +594,14 @@ func (i *Controller) nodeUpdater(obj runtime.Object, key string) error {
 	allocatedIp, err := i.annotateGlobalIp(key, existingGlobalIp)
 	if err != nil { // failed to get globalIp or failed to update, we want to retry
 		logAndRequeue(key, i.nodeWorkqueue)
-		return fmt.Errorf("failed to annotate GlobalIp to node %s: %v", key, err)
+		return fmt.Errorf("failed to annotate globalIp to node %s: %v", key, err)
```
Comment (mangelajo) [inline pkg/globalnet/controllers/ipam/ipam.go:597]:
```suggestion
		return fmt.Errorf("failed to annotate globalIp to node %q: %+v", key, err)
```

File: pkg/globalnet/controllers/ipam/ipam.go [hunk 23]
```diff
@@ -566,13 +594,14 @@ func (i *Controller) nodeUpdater(obj runtime.Object, key string) error {
...
 	}
 
 	// This case is hit in one of the two situations
 	// 1. when the Worker Node does not have the globalIp annotation and a new globalIp is allocated
 	// 2. when the current globalIp annotation on the Node does not match with the info maintained by ipPool
 	if allocatedIp != "" {
+		klog.V(log.DEBUG).Infof("Allocating globalIp %s to Node %s ", allocatedIp, key)
```
Comment (mangelajo) [inline pkg/globalnet/controllers/ipam/ipam.go:604]:
```suggestion
		klog.V(log.DEBUG).Infof("Allocating globalIp %s to Node %q ", allocatedIp, key)
```

File: pkg/globalnet/controllers/ipam/ipam.go [hunk 24]
```diff
@@ -592,7 +621,7 @@ func (i *Controller) nodeUpdater(obj runtime.Object, key string) error {
 			return err
 		}
 	} else if existingGlobalIp != "" {
-		klog.V(log.DEBUG).Infof("GatewayNode seems to have migrated, sync rules for node %q", node.Name)
+		klog.V(log.DEBUG).Infof("Node %q already has globalIp %q annotation, syncing rules", key, existingGlobalIp)
```
Comment (mangelajo) [inline pkg/globalnet/controllers/ipam/ipam.go:624]:
```suggestion
		klog.V(log.DEBUG).Infof("Node %q already has globalIp %s annotation, syncing rules", key, existingGlobalIp)
```

File: pkg/globalnet/controllers/ipam/gatewaymonitor.go [hunk 25]
```diff
@@ -121,6 +122,7 @@ func (i *GatewayMonitor) processNextEndpoint() bool {
 			return fmt.Errorf("error retrieving submariner endpoint object %s: %v", name, err)
 		}
 
+		klog.V(log.DEBUG).Infof("In processNextEndpoint, endpoint info: %v", endpoint)
```
Comment (mangelajo) [inline pkg/globalnet/controllers/ipam/gatewaymonitor.go:125]:
```suggestion
		klog.V(log.DEBUG).Infof("In processNextEndpoint, endpoint info: %+v", endpoint)
```

File: pkg/globalnet/controllers/ipam/gatewaymonitor.go (hunk 2 above)
Comment (sridhargaddam) [reply pkg/globalnet/controllers/ipam/gatewaymonitor.go:80]: Done

File: pkg/globalnet/controllers/ipam/gatewaymonitor.go (hunk 25 above)
Comment (sridhargaddam) [reply pkg/globalnet/controllers/ipam/gatewaymonitor.go:125]: Done

File: pkg/globalnet/controllers/ipam/ipam.go (hunk 3 above)
Comment (sridhargaddam) [reply pkg/globalnet/controllers/ipam/ipam.go:181]: haha...I'm convinced too ;-)

**Pull request merged**